x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/core/expression/UnresolvedAttribute.java

```diff
@@ -97,14 +97,18 @@ protected Attribute clone(
         NameId id,
         boolean synthetic
     ) {
-        // TODO: This looks like a bug; making clones should allow for changes.
-        return this;
-    }
+        // assertions, "should never happen"
+        if (dataType != null) {
+            throw new UnresolvedException("clone() with data type [" + dataType + "]", this);
```

Comment:
nit: Let's throw ISE if that shouldn't happen?

Maybe we could also skip this if assertions aren't on?